src/octoprint/plugins/virtual_printer/virtual.py

```diff
@@ -697,10 +697,24 @@ def _gcode_M117(self, data: str) -> None:
         # we'll just use this to echo a message, to allow playing around with pause triggers
         if self._echoOnM117:
             try:
-                self._send("echo:%s" % re.search(r"M117\s+(.*)", data).group(1))
+                result = re.search(r"M117\s+(.*)", data).group(1)
+                self._send(f"echo:{result}")
             except AttributeError:
                 self._send("echo:")
 
+    def _gcode_M118(self, data: str) -> None:
+        try:
+            parameter, result = re.search(r"M118 (\w{2}) (.*)", data).groups()
+            if "A1" in parameter.upper():
+                self._send(f"{result}")
+            elif "E1" in parameter.upper():
+                self._send(f"echo:{result}")
+            else:
+                self._logger.exception(f"{parameter} is not recognised")
+                raise Exception
+        except Exception:
+            self._logger.debug("Error sending back echo command M118")
```

Comment:
Either I misunderstand the docs, or the logic here is wrong.

First of all, `A1` and `E1` are optional, they do not have to be there. `M118 hello` should still work, it doesn't here. Then, `A1` is supposed to prefix the output with a `//`, it doesn't here.

How about this instead, this also adheres more strictly to the actual firmware behaviour and doesn't try to capture any two characters as parameter, but only those that are defined:

```suggestion
        match = re.search(r"M118 (?:(?P<parameter>A1|E1|Pn[012])\s)?(?P<text>.*)")
        if not match:
            self._send("Unrecognized command parameters for M118")
        
        text = match.groupdict().get("text")
        parameter = match.groupdict().get("parameter")
        
        if parameter == "A1":
            self._send(f"//{text}")
        elif parameter == "E1":
            self._send(f"echo:{text}")
        else:
            self._send(text)

```

See [here](https://regex101.com/r/tVDSjP/1) for some testing of that regex.

Cross checking this with [the firmware implementation](https://github.com/MarlinFirmware/Marlin/blob/42d30ed102308f0c5cb04b70dc55f9f37012a157/Marlin/src/gcode/host/M118.cpp#L35) looks like it matches in behaviour.

Reply:
that sounds good